devito/types/array.py

```diff
@@ -345,3 +346,217 @@ def dim(self):
     @property
     def array(self):
         return self._array
+
+
+class Bundle(ArrayBasic):
```

Comment:
So how different is it com[ared to a VectorFunction? Seem to abstract the same type of object.

Reply:
Initially I tried to use a VectorFunction, but I encountered several issues... I also tried to subclass it, but no luck. I honestly don't remember what the actual culprit was, but basically I would have had to subclass VectorFunction and override nearly all of its methods, and even then, such subclass would still appear as a user-provided object to the eyes of the compiler, which is not ideal

As an Array resembles a DiscreteFunction but it's not a DiscreteFunction, a Bundle resembles a VectorFunction but it's not a VectorFunction

I'm not entirely convinced this is the best course of action (so u may have a point), but this is where I landed after a non negligible amount of thinking and trial-and-error

Reply:
I guess the atmpy matrix internals are too much for it ok.